internal/webhook/external/quay.go

```diff
@@ -0,0 +1,151 @@
+package external
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	xhttp "github.com/akuity/kargo/internal/http"
+	"github.com/akuity/kargo/internal/io"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	quaySecretDataKey       = "secret"
+	quay                    = "quay"
+	quayWebhookBodyMaxBytes = 2 << 20 // 2MB
+)
+
+func init() {
+	registry.register(
+		quay,
+		webhookReceiverRegistration{
+			predicate: func(cfg kargoapi.WebhookReceiverConfig) bool {
+				return cfg.Quay != nil
+			},
+			factory: newQuayWebhookReceiver,
+		},
+	)
+}
+
+// quayWebhookReceiver is an implementation of WebhookReceiver that handles
+// inbound webhooks from Quay.
+type quayWebhookReceiver struct {
+	*baseWebhookReceiver
+}
+
+// newQuayWebhookReceiver returns a new instance of quayWebhookReceiver.
+func newQuayWebhookReceiver(
+	c client.Client,
+	project string,
+	cfg kargoapi.WebhookReceiverConfig,
+) WebhookReceiver {
+	return &quayWebhookReceiver{
+		baseWebhookReceiver: &baseWebhookReceiver{
+			client:     c,
+			project:    project,
+			secretName: cfg.Quay.SecretRef.Name,
+		},
+	}
+}
+
+// GetDetails implements WebhookReceiver.
+func (q *quayWebhookReceiver) getReceiverType() string {
+	return quay
+}
+
+// getSecretValues implements WebhookReceiver.
+func (q *quayWebhookReceiver) getSecretValues(
+	secretData map[string][]byte,
+) ([]string, error) {
+	secretValue, ok := secretData[quaySecretDataKey]
+	if !ok {
+		return nil,
+			errors.New("Secret data is not valid for a Quay WebhookReceiver")
+	}
+	return []string{string(secretValue)}, nil
+}
+
+// GetHandler implements WebhookReceiver.
+func (q *quayWebhookReceiver) GetHandler() http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("identifying source repository")
+
```

Comment:
In many cases where the body is too large, we can spare ourselves the effort of the io.LimitRead. (However, we cannot trust that the `Content-Length` is always set, so falling back on the io.LimitRead is still imporant.)

```suggestion

		// Early check of Content-Length if available
		if contentLength := r.ContentLength; contentLength > githubWebhookBodyMaxBytes {
			xhttp.WriteErrorJSON(
				w,
				xhttp.Error(
					fmt.Errorf("content exceeds limit of %d bytes", githubWebhookBodyMaxBytes),
					http.StatusRequestEntityTooLarge,
				),
			)
			return
		}

```